src/kafka-extensions/registry.ts

```diff
@@ -0,0 +1,126 @@
+    }
+
+    private async getProcessor(): Promise<ClusterProviderProcessor> {
+        if (this.processor) {
+            return this.processor;
+        }
+        return vscode.commands.executeCommand(this.definition.commandId)
+            .then(result => {
+                return this.processor = <ClusterProviderProcessor>result;
+            });
+    }
+
+}
+
+const defaultClusterProviderId = 'default';
+
+const providers: Map<string, ClusterProvider> = new Map();
+
+export function getClusterProvider(clusterProviderId?: string): ClusterProvider | undefined {
+    intializeIfNeeded();
+    return providers.get(clusterProviderId || defaultClusterProviderId);
+}
+
+export function getClusterProviders(): ClusterProvider[] {
+    intializeIfNeeded();
+    return [...providers.values()];
+}
+
+function intializeIfNeeded() {
+    if (providers.size === 0) {
+        const all = collectClusterProviderDefinitions(vscode.extensions.all);
+        all.forEach(definition => {
+            providers.set(definition.id, new ClusterProvider(definition));
+        });
+    }
+}
+
+export interface ClusterProviderDefinition {
+    id: string;
+    name?: string;
+    commandId: string;
+}
+
+let existingExtensions: Array<ClusterProviderDefinition>;
+
+export function collectClusterProviderDefinitions(extensions: readonly vscode.Extension<any>[]): ClusterProviderDefinition[] {
+    const result = [] as Array<ClusterProviderDefinition>;
+    if (extensions && extensions.length) {
+        for (const extension of extensions) {
+            const contributesSection = extension.packageJSON['contributes'];
+            if (contributesSection) {
+                const kafkaExtension = contributesSection['kafka'];
+                if (kafkaExtension) {
+                    const clusterProviders = kafkaExtension['clusterProviders'];
+                    if (Array.isArray(clusterProviders) && clusterProviders.length) {
+                        for (const provider of clusterProviders) {
+                            result.push(provider as ClusterProviderDefinition);
+                        }
+                    }
+                }
+            }
+        }
+    }
+    // Make a copy of extensions:
+    existingExtensions = result.slice();
+    return result;
+}
+
+export function onExtensionChange(extensions: readonly vscode.Extension<any>[]) {
+    if (!existingExtensions) {
+        return;
+    }
+    const oldExtensions = new Set(existingExtensions.slice());
+    const newExtensions = collectClusterProviderDefinitions(extensions);
+    let hasChanged = (oldExtensions.size !== newExtensions.length);
+    if (!hasChanged) {
+        for (const newExtension of newExtensions) {
+            if (!oldExtensions.has(newExtension)) {
+                hasChanged = true;
+                break;
+            }
+        }
+    }
+}
+
+export function registerClusterProcessorCommand() {
+    vscode.commands.registerCommand('vscode-kafka.cluster.providers', () => {

```

Comment:
need to call context.subscriptions.push()

Reply:
This code was removed since we don't use command anymore.